File: src/main/java/cote/programmers/PresentByN.java
package cote.programmers;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

class PresentByN {
    // dp에는 숫자 N으로 만들어진 경우의 수가 들어간다.
    public int solution(int N, int number) {
        int answer = -1;
        List<Set<Integer>> dp = new ArrayList<>();

        // 숫자 N으로 만들어진 dp 리스트 만들기
        for (int i = 1; i <= 8; i++) {
            Set<Integer> tmp = new HashSet<>();
            tmp.add(Integer.parseInt(Integer.toString(N).repeat(i)));
            dp.add(tmp);
        }


        return answer;
    }
}

File: cote-py/programmers/PresentByN.py
def solution(N, number):
    answer = -1
    dp = [set([int(str(N)*i)]) for i in range(1, 9)]

    # 8번 반복
    for i in range(8):
        for j in range(i):
            for first in dp[j]:
                for second in dp[i-j-1]:
                    dp[i].add(first + second)
                    dp[i].add(first - second)
                    dp[i].add(first - second)
                    if second != 0:
                        dp[i].add(first // second)
        if number in dp[i]:
            return i+1

    return answer


print(solution(5, 12))
print(solution(2, 11))
